lib/semantics/expression.cc

```diff
@@ -2439,6 +2532,68 @@ std::optional<ActualArgument> ArgumentAnalyzer::Analyze(
   }
 }
 
+bool ArgumentAnalyzer::AreConformable() const {
+  CHECK(!fatalErrors_ && actuals_.size() == 2);
+  return evaluate::AreConformable(*actuals_[0], *actuals_[1]);
+}
+
+const Symbol *ArgumentAnalyzer::FindDefinedOp(const char *opr) {
+  const auto &scope{context_.context().FindScope(source_)};
+  return scope.FindSymbol(parser::CharBlock{"operator("s + opr + ')'});
+}
+
+std::optional<DynamicType> ArgumentAnalyzer::GetType(std::size_t i) const {
+  return i < actuals_.size() ? actuals_[i].value().GetType() : std::nullopt;
+}
+
+// Report error resolving opr when there is a user-defined one available
+void ArgumentAnalyzer::SayNoMatch(const char *opr) {
+  auto rank0{actuals_[0]->Rank()};
+  if (actuals_.size() == 1) {
+    if (rank0 > 0) {
+      context_.Say("No user-defined or intrinsic %s operator matches "
+                   "rank %d array of %s"_err_en_US,
+          ToUpperCase(opr), rank0, TypeAsFortran(0));
+    } else {
+      context_.Say("No user-defined or intrinsic %s operator matches "
+                   "operand type %s"_err_en_US,
+          ToUpperCase(opr), TypeAsFortran(0));
+    }
+  } else {
+    auto rank1{actuals_[1]->Rank()};
+    if (rank0 > 0 && rank1 > 0 && rank0 != rank1) {
+      context_.Say("No user-defined or intrinsic %s operator matches "
+                   "rank %d array of %s and rank %d array of %s"_err_en_US,
+          ToUpperCase(opr), rank0, TypeAsFortran(0), rank1, TypeAsFortran(1));
+    } else {
+      context_.Say("No user-defined or intrinsic %s operator matches "
+                   "operand types %s and %s"_err_en_US,
+          ToUpperCase(opr), TypeAsFortran(0), TypeAsFortran(1));
+    }
+  }
+}
+
+std::string ArgumentAnalyzer::TypeAsFortran(std::size_t i) {
+  if (std::optional<DynamicType> type{GetType(i)}) {
+    return type->category() == TypeCategory::Derived
+        ? "TYPE("s + type->AsFortran() + ')'
+        : type->category() == TypeCategory::Character
+            ? "CHARACTER(KIND="s + std::to_string(type->kind()) + ')'
+            : ToUpperCase(type->AsFortran());
+  } else {
+    return "untyped"s;
+  }
+}
+
+bool ArgumentAnalyzer::AnyUntypedOperand() {
+  for (const auto &actual : actuals_) {
```

Comment:
Might be more natural as a `std::find()`, might not.

Reply:
It is a little shorter, but more complex, so I left it as-is.